File: python_project/algorithms/dda.py
"""
DDA (Digital Differential Analyzer) Line Drawing Algorithm.
Generates step-by-step states for coordinate grid visualization.
"""

class DDALine:
    def __init__(self, x1: int, y1: int, x2: int, y2: int):
        self.x1 = int(x1)
        self.y1 = int(y1)
        self.x2 = int(x2)
        self.y2 = int(y2)

    def generate_steps(self):
        """
        Calculates line points using the DDA formula:
        dx = x2 - x1
        dy = y2 - y1
        steps = max(|dx|, |dy|)
        x_inc = dx / steps
        y_inc = dy / steps
        """
        dx = self.x2 - self.x1
        dy = self.y2 - self.y1
        steps = max(abs(dx), abs(dy))

        if steps == 0:
            return [{
                "step": 0,
                "total_steps": 1,
                "x_float": self.x1,
                "y_float": self.y1,
                "pixel": (self.x1, self.y1),
                "dx": 0,
                "dy": 0,
                "x_inc": 0,
                "y_inc": 0,
                "explanation": f"Start and end points coincide at ({self.x1}, {self.y1}).",
                "pixels": [(self.x1, self.y1)]
            }]

        x_inc = dx / steps
        y_inc = dy / steps

        cur_x = float(self.x1)
        cur_y = float(self.y1)
        step_records = []
        accumulated_pixels = []

        for i in range(steps + 1):
            px = round(cur_x)
            py = round(cur_y)
            accumulated_pixels.append((px, py))

            if i == 0:
                expl = f"Initial pixel ({px}, {py}). Total steps = {steps}, x_inc = {x_inc:.3f}, y_inc = {y_inc:.3f}"
            elif i == steps:
                expl = f"Endpoint ({px}, {py}) reached. Rasterization complete."
            else:
                expl = f"Step {i}: (x={cur_x:.2f}, y={cur_y:.2f}) rounded to ({px}, {py})"

            step_records.append({
                "step": i,
                "total_steps": steps,
                "x_float": round(cur_x, 3),
                "y_float": round(cur_y, 3),
                "pixel": (px, py),
                "dx": dx,
                "dy": dy,
                "x_inc": round(x_inc, 4),
                "y_inc": round(y_inc, 4),
                "explanation": expl,
                "pixels": list(accumulated_pixels)
            })

            cur_x += x_inc
            cur_y += y_inc

        return step_records

File: src/algorithms/dda.ts
import { DDAStepState, PixelCoord } from '../types';

export function generateDDASteps(x1: number, y1: number, x2: number, y2: number): DDAStepState[] {
  const steps: DDAStepState[] = [];
  const dx = x2 - x1;
  const dy = y2 - y1;
  const numSteps = Math.max(Math.abs(dx), Math.abs(dy));

  if (numSteps === 0) {
    const px: PixelCoord = { x: Math.round(x1), y: Math.round(y1) };
    return [{
      step: 0,
      totalSteps: 1,
      xFloat: x1,
      yFloat: y1,
      xPixel: px.x,
      yPixel: px.y,
      dx: 0,
      dy: 0,
      xInc: 0,
      yInc: 0,
      slope: 'Undefined (point)',
      explanation: `Start and end points coincide at (${x1}, ${y1}). Single pixel drawn.`,
      historyPixels: [px]
    }];
  }

  const xInc = dx / numSteps;
  const yInc = dy / numSteps;
  const slope = dx === 0 ? 'Infinity (Vertical)' : (dy / dx).toFixed(4);

  let curX = x1;
  let curY = y1;
  const history: PixelCoord[] = [];

  for (let i = 0; i <= numSteps; i++) {
    const xPix = Math.round(curX);
    const yPix = Math.round(curY);
    history.push({ x: xPix, y: yPix, label: `(${xPix},${yPix})` });

    let explanation = '';
    if (i === 0) {
      explanation = `Initial point plotted at (${xPix}, ${yPix}). Total steps needed = max(|dx|, |dy|) = max(${Math.abs(dx)}, ${Math.abs(dy)}) = ${numSteps}. Increments: x_inc = ${xInc.toFixed(3)}, y_inc = ${yInc.toFixed(3)}.`;
    } else if (i === numSteps) {
      explanation = `Final endpoint reached at (${xPix}, ${yPix}). Line rasterization complete.`;
    } else {
      explanation = `Step ${i}: Float values (x = ${curX.toFixed(2)}, y = ${curY.toFixed(2)}) rounded to nearest pixel raster coordinates (${xPix}, ${yPix}).`;
    }

    steps.push({
      step: i,
      totalSteps: numSteps,
      xFloat: Number(curX.toFixed(3)),
      yFloat: Number(curY.toFixed(3)),
      xPixel: xPix,
      yPixel: yPix,
      dx,
      dy,
      xInc: Number(xInc.toFixed(4)),
      yInc: Number(yInc.toFixed(4)),
      slope,
      explanation,
      historyPixels: [...history]
    });

    curX += xInc;
    curY += yInc;
  }

  return steps;
}
